File: cpp/GCD_Problem.py
import math
def build(given, tree, low, high, pos):
    if(low == high):
        tree[pos][0] = given[low]
        tree[pos][1] = 0
        return tree[pos]
    mid = low + (high-low)//2
    left = build(given, tree, low, mid, 2*pos + 1)
    right = build(given, tree, mid+1, high, 2*pos + 2)
    tree[pos][0] = math.gcd(left[0], right[0])
    if(left[1] == right[1] == 0):
        tree[pos][1] = math.gcd(left[0], right[0])
    elif(left[1] == 0):
        tree[pos][1] = right[1]
    elif(right[1] == 0):
        tree[pos][1] = left[1]
    else:
        tree[pos][1] = left[1] + right[1]
    return tree[pos]
def query(given, tree, qlow, qhigh, low, high, pos):
    pass
    if(qlow > high or qhigh <low):
        return 0
    if(qlow <= low and qhigh >= high):
        return tree[pos]
    mid = low + (high - low)//2
    return []

N,M = map(int, input().split())
tree = [[0 for i in range(2)] for j in range(4*N)]
given = list(map(int, input().split()))
build(given, tree, 0, N-1, 0)
print(tree)

File: cpp/GCD_Problem.cpp
#include <bits/stdc++.h>
#define lli long long int
#define endl "\n"
using namespace std;
lli sum =0;
lli build(lli given[], lli segTree[], lli low, lli high, lli pos)
{
    if (low == high)
    {
        segTree[pos] = given[low];
        return segTree[pos];
    }
    lli mid = low + (high - low) / 2;
    segTree[pos] = __gcd(build(given, segTree, low, mid, 2 * pos + 1), build(given, segTree, mid + 1, high, 2 * pos + 2));
    return segTree[pos];
}
lli query(lli given[], lli segTree[],lli qlow,lli qhigh ,lli low, lli high,lli pos){
    // cout<<low<<" "<<high<<" "<<pos<<endl;
    if(qlow > high or qhigh < low){
        return 0;
    }
    if(qlow <= low and qhigh >= high){
        return segTree[pos];
    }
    lli mid = low + (high - low)/2;
    return __gcd(query(given, segTree, qlow, qhigh, low, mid, 2*pos + 1), query(given, segTree, qlow, qhigh, mid+1, high, 2*pos +2));

}
int main(int argc, char const *argv[])
{
    lli N, M;
    cin >> N >> M;
    lli given[N];
    lli segTree[4 * N + 1] = {0};
    for (int i = 0; i < N; i++)
    {
        cin >> given[i];
    }
    build(given, segTree, 0, N - 1, 0);
    for(lli i = 0;i<M;i++){
        lli low, high;
        cin>>low>>high;
        low--;
        high--;
        lli sum = 0;
        for(int j = low;j<= high;j++){
            for(int k = j;k<= high;k++){
                sum += query(given, segTree, j, k, 0, N-1, 0);
            }
        }
        cout<<sum<<endl;
    }
    return 0;
}
